### DISCUSSION TITLE
Nordic UARTService in.. gasp.. Java?

### DISCUSSION THREAD

**Discussion by lebenski** [pybricks/support#1465] [General]
Hello,

I'm new to the world of Pybricks and Lego robotics in general but I've got some cool stuff up and running and am looking to take it a step further.  I'm looking to communicate with my hubs from my PC and I see that the way to do that is via Nordic UART Service.  I really need to do this in Java for reasons of convenience with my related projects however I've only found a couple of reference implementations which are very old and bring in a whole bunch of Android Studio baggage with them.

https://github.com/NordicPlayground/Android-nRF-UART/blob/master/app/src/main/java/com/nordicsemi/nrfUARTv2/UartService.java
https://github.com/adafruit/Adafruit_Android_BLE_UART/blob/master/app/src/main/java/com/adafruit/bleuart/BluetoothLeUart.java

My questions are.. is Android a necessary dependency to set up communication with the Hub via UART?  And is anyone aware of a lightweight (few dependencies) sample implementation in Java that I could reference?

Many Thanks,
Ben

**Comment by laurensvalk** [comment]
> is Android a necessary dependency to set up communication with the Hub via UART

Android should not not needed per se (but maybe Android projects can be useful inspiration).

You'd just need some sort of library or interface with Bluetooth Low Energy. Here is some Python example code: https://pybricks.com/projects/tutorials/wireless/hub-to-device/pc-communication/

**Reply by lebenski** [reply]
Thank you!  I got that up and working great.  After a lot of research it seems like a BLE Java Library that can operate on MacOS simply does not exist yet.  I may have to delve deeper to understand more so that I can write one myself.  If anyone else has found something that I haven't then please let me know.

As an aside I noticed Issue 1469 that you create this morning.  The latest code is no longer using NUS?  What is the new setup for PC to Hub comms?

**Reply by laurensvalk** [reply]
Thanks, and good catch. You’re already using the correct code, so you can use the same principles (in any language with a ble framework). We just need to revise the description to no longer say we use NUS as we used to.

**Reply by lebenski** [reply]
Thanks again.  I've got this up and working but I just realised that it seems to have nuked the hub to hub communication that I had going on using this method: https://pybricks.com/projects/tutorials/wireless/hub-to-hub/broadcast/

I have a City Hub (Hub A) mounted on a train and driving a DCMotor which is observing messages on the broadcast channel from another City Hub (Hub B) connected to a Color Sensor.  When I connect my PC to Hub A via this BleakClient & the UArtService, Hub A seems to lose a lot of the broadcast messages that it would otherwise receive from Hub B if it weren't connected to the PC.

Presumably the problem is due to the Known Issue at the bottom of the Hub to Hub communication page? (Known issue: Slow communication while connected to computer).

Is there any workaround for this?  I'm not limited to using the CityHub if that helps.  Do any of the other hubs do a better job of listening on both the UArtService connection and in the inter-hub broadcast channel at the same time?

I appreciate your help!

**Reply by laurensvalk** [reply]
The broadcasting is less affected by this, so maybe you can change your setup to take this into account?

Indeed, observing and being connected at the same time seems to slow it down.

It also works better on Spike and Mindstorms robot inventor.

**Reply by lebenski** [reply]
Thanks for your help again!  It turned out that the broadcasting/observing was actually working fine when one of the hubs was connected to my PC via the UART Service.  In my lack of experience with Python I simply hadn't realised that stdin.buffer.read(3) was a blocking call!  So it meant that my hub was busy blocking there waiting for another command from the PC and not able to observe the channel... duh!  I changed it to this and we're all good now.

```
    list = uselect.select([stdin], [], [], 0.01)
    if list[0]:
        cmd = stdin.buffer.read(3)
    else:
        cmd = None
```

Perhaps there is a better solution but this seemed to work fine.

Anyway back to the original topic of this discussion.. I think this Bleak app is going to work fine for me.  It's a shame that there isn't a readymade BLE library for Java on MacOS, but i'll just get my Java app talking to this Bleak app and it'll work just fine!

Thanks again,
Ben
